runtime/src/accounts_db.rs

```diff
@@ -1381,9 +1416,68 @@ impl AccountsDB {
             .and_then(|account_storage_entry| account_storage_entry.get_account(account_info))
     }
 
-    fn find_storage_candidate(&self, slot: Slot) -> Arc<AccountStorageEntry> {
+    fn try_recycle_and_insert_store(
+        &self,
+        slot: Slot,
+        min_size: u64,
+        max_size: u64,
+    ) -> Option<Arc<AccountStorageEntry>> {
+        let store = self.try_recycle_store(slot, min_size, max_size)?;
+        self.insert_store(slot, store.clone());
+        Some(store)
+    }
+
+    fn try_recycle_store(
+        &self,
+        slot: Slot,
+        min_size: u64,
+        max_size: u64,
+    ) -> Option<Arc<AccountStorageEntry>> {
+        let mut max = 0;
+        let mut min = std::u64::MAX;
+        let mut avail = 0;
+        let mut recycle_stores = self.recycle_stores.write().unwrap();
+        for (i, store) in recycle_stores.iter().enumerate() {
```

Comment:
If `recycle_stores` was a `BTreeMap<usize, Vec<Arc<AccountStorageEntry>>` keyed by the store size, then we would only need a lookup and a `pop()` and can avoid this scan under the lock.

Reply:
When I look at the perf, the current implementation is quite fast. I will experiment with this in a follow-up.